### PR TITLE
Add "Button for removing items from profile pages" addon

### LINKED ISSUE #5958: On the favorite projects, following, etc. pages, add an option to remove the item from that page
### Why this would be helpful

Sometimes, you want to clean up your shared or favorites projects, for example. You also don't want to open every single project in a new tab.

### How this addon works

This addon adds an X button to every item of almost every page of rows on your profile page that does the following:
| Page | What the X does |
|--------|--------|
| Shared Projects | Unshare the project |
| Favorite Projects | Unfavorite the project |
| Studios I'm Following | Unfollow the studio |
| Studios I Curate | Leave the studio |
| Following | Unfollow the user | 

### Possible alternatives

* A way to open the user/project without leaving the page. Getting this right would be more difficult than just adding this undo option, though.
* This could be merged with `scratchstats` for a "Better Profiles" addon, but these addons are pretty unrelated.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mybearworld**
Resolves #5958

### Changes

Added "Button for removing items from profile pages" addon (internally `remove-from-profile-pages`) that adds buttons that remove certain projects/user/studios from the page.

### Reason for changes

Sometimes, you want to clean up your shared or favorites projects, for example. You also don't want to open every single project in a new tab.

### Tests

Tested on Edge 114.0.1793.0 (Official build) dev (64-bit).

### CODE REVIEW

**Comment by mybearworld** [comment]
I'm not sure what I did wrong for the Prettier check to fail...

**Comment by rgantzos** [comment]
@mybearworld You don't need to update the translation files; it's already done by the system that translates it with Transifex

If prettier fails, it means that the code is formatted incorrectly. I'm assuming there will be changes to this code before it's merged (and it looks pretty good right now), so it's not something you need to worry about too much right now, but you might want to just put it into a [prettier formatter](https://prettier.io/playground/).

**Comment by mybearworld** [comment]
> @mybearworld You don't need to update the translation files; it's already done by the system that translates it with Transifex

How else would it know what text to use?

> If prettier fails, it means that the code is formatted incorrectly. I'm assuming there will be changes to this code before it's merged (and it looks pretty good right now), so it's not something you need to worry about too much right now, but you might want to just put it into a [prettier formatter](https://prettier.io/playground/).

I have formatted the code with Prettier on VSCode already (with the .prettierrc.json in the repository).

**Comment by rgantzos** [comment]
> > @mybearworld You don't need to update the translation files; it's already done by the system that translates it with Transifex
> 
> How else would it know what text to use?
> 
Sorry my bad, I misread it.

**Comment by mybearworld** [comment]
The error `fatal: You are not currently on a branch.` doesn't really sound like a problem with the formatting to me - is it just a weird error message?

**Comment by WorldLanguages** [comment]
Don't worry about Prettier

**Comment by mybearworld** [comment]
> Don't worry about Prettier

Okay

**Comment by mybearworld** [comment]
I'll implement dynamic setting changes as well soon.

**Comment by mybearworld** [comment]
Is this too many settings?

Nine settings.

**Comment by Secret-chest** [comment]
Dedicated pages only would be very useful.

**Comment by BroJac5246** [comment]
> Is this too many settings?
> 
> Nine settings.

It's a lot, but they're all useful settings, so it's _probably_ okay.

**Comment by DNin01** [comment]
For user-experience reasons, I don't know if every option has to stay, but I still have to review.

**Comment by DNin01** [comment]
I sure wish we could [https://github.com/ScratchAddons/ScratchAddons/issues/5719](better organize these settings).

File: addons/remove-from-profile-rows/addon.json [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+      "matches": ["profiles", "https://scratch.mit.edu/users/*/*"]
+    }
+  ],
+  "settings": [
+    {
+      "name": "Only show on the dedicated pages",
+      "id": "dedicated_pages",
+      "type": "boolean",
+      "default": false
+    },
```
Comment (DNin01) [inline addons/remove-from-profile-rows/addon.json:25]: If this means only show the button to remove items on full lists and not profile pages, then I don't think this option is necessary. Instead, I think the buttons should always only appear on dedicated pages. Profile pages are more likely used for browsing and not editing.

File: addons/remove-from-profile-rows/addon.json [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    {
+      "name": "Only show the button on hover",
```
Comment (DNin01) [inline addons/remove-from-profile-rows/addon.json:27]: Could probably be renamed:
```suggestion
      "name": "Only show remove button while hovering over items",
```

File: addons/remove-from-profile-rows/addon.json [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    },
+    {
+      "name": "Only show confirmation on destructive actions (Unsharing, leaving studios)",
+      "id": "only_destructive",
+      "type": "boolean",
+      "default": true,
+      "if": {
+        "settings": {"show_confirmation": true}
+      }
+    },
```
Comment (DNin01) [inline addons/remove-from-profile-rows/addon.json:46]: I don't know about you, but I feel these two settings could be improved. The second setting here _removes_ functionality when _enabled_, and it's also slightly unclear what this should do if disabled - it seems to not imply by itself that when disabled, a confirmation will appear for all actions.

My idea to fix this problem would be to replace these on/off switches with a multi-choice setting named "confirmation before removing" with three choices, "off", "before unsharing projects and leaving studios" and "all actions", or something else.

**Review by DNin01 — COMMENTED** [review]
I've yet to test the addon, but here are some suggestions for its settings.

File: addons/remove-from-profile-rows/addon.json (hunk 3 above)
Comment (mybearworld) [reply addons/remove-from-profile-rows/addon.json:46]: That does sound a lot better, and it also cuts down on the amount of settings required.

Comment (mybearworld) [reply addons/remove-from-profile-rows/addon.json:46]: I did that, but now it looks like this:
image
And that isn't really good.

Comment (Secret-chest) [reply addons/remove-from-profile-rows/addon.json:46]: The three-way setting could be first.

Comment (mybearworld) [reply addons/remove-from-profile-rows/addon.json:46]: image
That's better, but I don't think it's very ideal.

**Review by DNin01 — COMMENTED** [review]
We still need to work on Dynamic Disable for this addon.
Xs placed over images

Luckily, 2.0 → 3.0 doesn't break anything. Nice!

What is slightly concerning is accidentally clicking the X, especially on profile pictures which are already small. Maybe it could be placed _on_ the corner of the thumbnails, like in the Scratch-2.0 studio design, further out of the way. It might also be cool if it looked like the delete button on the sprites list from the editor.

**Comment by BroJac5246** [comment]
> We still need to work on Dynamic Disable for this addon. Xs placed over images
> 
> Luckily, 2.0 → 3.0 doesn't break anything. Nice!
> 
> What is slightly concerning is accidentally clicking the X, especially on profile pictures which are already small. Maybe it could be placed _on_ the corner of the thumbnails, like in the Scratch-2.0 studio design, further out of the way. It might also be cool if it looked like the delete button on the sprites list from the editor.

On the thumbnails' corners would probably be the better placement because it's a lot more efficient.

**Comment by mybearworld** [comment]
> Maybe it could be placed _on_ the corner of the thumbnails, like in the Scratch-2.0 studio design, further out of the way.

> On the thumbnails' corners would probably be the better placement because it's a lot more efficient.

It already is on the corner of the thumbnails - your screenshot seems to be an issue with CSS not getting applied for some reason.
image


> It might also be cool if it looked like the delete button on the sprites list from the editor.

That would only work for when `scratchr2` is enabled, otherwise it would feel out of place.

**Comment by Secret-chest** [comment]
What about a setting to only show it on hover?

**Comment by Secret-chest** [comment]
The button should be round and have a proper icon, not an `X`.

**Comment by mybearworld** [comment]
Dynamic disable is fixed now - I accidentally used `() => disable` instead of just `disable`.

> What about a setting to only show it on hover?

Already exists.

> The button should be round and have a proper icon, not an `X`.

I'm using a simple <button> element, so it has the same style as all other buttons, I thought that would make the most sense.

**Comment by mybearworld** [comment]
Confirmation messages could probably use some more detail (e.g. "Unfavorite Sample Project?" instead of "Remove this item from this row?")

File: addons/remove-from-profile-rows/addon.json (hunk 3 above)
Comment (mybearworld) [reply addons/remove-from-profile-rows/addon.json:46]: I think when confirmations are disabled, it's expected that you won't get any confirmations.

**Comment by BroJac5246** [comment]
> > The button should be round and have a proper icon, not an `X`.
> 
> I'm using a simple `<button>` element, so it has the same style as all other buttons, I thought that would make the most sense.

For consistency, it might make sense to use what is used in studios, including based on if the studio is 2.0- or 3.0-styled.

**Comment by mybearworld** [comment]
> > > The button should be round and have a proper icon, not an `X`.
> > 
> > 
> > I'm using a simple `<button>` element, so it has the same style as all other buttons, I thought that would make the most sense.
> 
> For consistency, it might make sense to use what is used in studios, including based on if the studio is 2.0- or 3.0-styled.

That would make sense.

**Comment by BroJac5246** [comment]
> > > > The button should be round and have a proper icon, not an `X`.
> > > 
> > > 
> > > I'm using a simple `<button>` element, so it has the same style as all other buttons, I thought that would make the most sense.
> > 
> > For consistency, it might make sense to use what is used in studios, including based on if the studio is 2.0- or 3.0-styled.
> 
> There's not really a remove button on studios.

I mean the buttons to remove a project from a studio while within a studio, sorry.

**Comment by mybearworld** [comment]
> > > > > The button should be round and have a proper icon, not an `X`.
> > > > 
> > > > 
> > > > I'm using a simple `<button>` element, so it has the same style as all other buttons, I thought that would make the most sense.
> > > 
> > > 
> > > For consistency, it might make sense to use what is used in studios, including based on if the studio is 2.0- or 3.0-styled.
> > 
> > 
> > There's not really a remove button on studios.
> 
> I mean the buttons to remove a project from a studio while within a studio, sorry.

Again, there isn't really a button for that - just three dots with an option to remove there. This seems like too much for an addon like this.

**Comment by BroJac5246** [comment]
> > > > > > The button should be round and have a proper icon, not an `X`.
> > > > > 
> > > > > 
> > > > > I'm using a simple `<button>` element, so it has the same style as all other buttons, I thought that would make the most sense.
> > > > 
> > > > 
> > > > For consistency, it might make sense to use what is used in studios, including based on if the studio is 2.0- or 3.0-styled.
> > > 
> > > 
> > > There's not really a remove button on studios.
> > 
> > 
> > I mean the buttons to remove a project from a studio while within a studio, sorry.
> 
> Again, there isn't really a button for that - just three dots with an option to remove there. This seems like too much for an addon like this.

I just realized I haven't actually removed a project from a studio since they got changed to scratch-www 🤦
I was thinking more along the lines of the 2.0 studios, where there was just a circular "x" button in the top right (I believe). If the 3.0-styled pages addon is enabled then we can change the button to be blue to better fit the 3.0 theme.

**Comment by mybearworld** [comment]
> I was thinking more along the lines of the 2.0 studios, where there was just a circular "x" button in the top right (I believe). If the 3.0-styled pages addon is enabled then we can change the button to be blue to better fit the 3.0 theme.

The changing button to blue part is automatically done with the `scratchr2` addon already. It being circular and on the top right would work, but I don't think it's too necessary.

**Comment by mybearworld** [comment]
Error messages are now more detailed:
scratch.mit.edu says: Are you sure you want to leave "Untitled Studio"?
scratch.mit.edu says: Failed leaving "Untitled Studio". This could be because the studio doesn't exist anymore, you own the studio, or you're signed out.

**Comment by DNin01** [comment]
> > Maybe it could be placed _on_ the corner of the thumbnails, like in the Scratch-2.0 studio design, further out of the way.
> 
> 
> 
> It already is on the corner of the thumbnails - your screenshot only happened because CSS isn't getting applied due to the problems with dynamic disable (that should be an easy fix) (it was).

By "_on_ the corner" I meant placed on top of the border, half on top of the image and half outside the border. Right now it's inside the image.

**Comment by mybearworld** [comment]
> > > Maybe it could be placed _on_ the corner of the thumbnails, like in the Scratch-2.0 studio design, further out of the way.
> > 
> > 
> > It already is on the corner of the thumbnails - your screenshot only happened because CSS isn't getting applied due to the problems with dynamic disable (that should be an easy fix) (it was).
> 
> By "_on_ the corner" I meant placed on top of the border, half on top of the image and half outside the border. Right now it's inside the image.

Like this, but somehow with the border not disappearing?
image

**Comment by BroJac5246** [comment]
> > > > Maybe it could be placed _on_ the corner of the thumbnails, like in the Scratch-2.0 studio design, further out of the way.
> > > 
> > > 
> > > It already is on the corner of the thumbnails - your screenshot only happened because CSS isn't getting applied due to the problems with dynamic disable (that should be an easy fix) (it was).
> > 
> > By "_on_ the corner" I meant placed on top of the border, half on top of the image and half outside the border. Right now it's inside the image.
> 
> Like this, but somehow with the border not disappearing?
> image
> 

Much better!

**Comment by mybearworld** [comment]
> > > > > Maybe it could be placed _on_ the corner of the thumbnails, like in the Scratch-2.0 studio design, further out of the way.
> > > > 
> > > > 
> > > > It already is on the corner of the thumbnails - your screenshot only happened because CSS isn't getting applied due to the problems with dynamic disable (that should be an easy fix) (it was).
> > > 
> > > 
> > > By "_on_ the corner" I meant placed on top of the border, half on top of the image and half outside the border. Right now it's inside the image.
> > 
> > 
> > Like this, but somehow with the border not disappearing?
> 
> Much better!

I think so too! But I can't figure out how to have the border not disappear, because it goes beyond the `item`.

**Comment by mybearworld** [comment]
The only way I can think of is having a margin at the top of the items - but this looks weird.
image

**Comment by DNin01** [comment]
> I think so too! But I can't figure out how to have the border not disappear, because it goes beyond the `item`.

This doesn't seem to happen on the dedicated pages, which is one advantage of only supporting dedicated pages, as I suggested earlier.

**Comment by mybearworld** [comment]
> > I think so too! But I can't figure out how to have the border not disappear, because it goes beyond the `item`.
> 
> This doesn't seem to happen on the dedicated pages, which is one advantage of only supporting dedicated pages, as I suggested earlier.

That probably does make sense - it'd have to have a different name, though.

**Comment by mybearworld** [comment]
Updated settings page
This is much better.

**Comment by mybearworld** [comment]
I didn't know renaming the branch closed the issue...

File: addons/remove-from-profile-rows/addon.json (hunk 3 above)
Comment (mybearworld) [reply addons/remove-from-profile-rows/addon.json]: Done - https://github.com/ScratchAddons/ScratchAddons/pull/5960#issuecomment-1528868196

File: addons/remove-from-profile-rows/addon.json (hunk 1 above)
Comment (mybearworld) [reply addons/remove-from-profile-rows/addon.json]: Done with https://github.com/ScratchAddons/ScratchAddons/pull/5960/commits/8fc107c0f36b3b74dff9e769cb11ae1d6d900f6c.

**Pull request closed without merging**